datalad/distribution/dataset.py

```diff
@@ -218,8 +219,10 @@ def repo(self):
                     lgr.log(5,
                             "Oops -- guess on repo type was wrong?: %s",
                             exc_str(exc))
-                    pass
+                    self._repo = None
```

Comment:
> eh, and we really can't do regular lazy evaluation here?

We never did. Run that same command on master and watch the exact same thing happening.

Reply:
Re 1 and 2: Agree.

Reply:
This also means, that this:

> (so previous value of ._repo gets entirely forgotten/overwritten)

is not correct. It would be the case if we set it to `None` before. But since we don't, the line
```python
self._repo = cls(self._path, create=False, **kw)
```
will return the exact same object we started with, if nothing changed.